org/mozilla/jss/pkcs11/attrs/CKAttribute.c

```diff
@@ -0,0 +1,271 @@
+#include <nspr.h>
+#include <nss.h>
+#include <pkcs11t.h>
+#include <jni.h>
+
+#include "CKAttribute.h"
+#include "StaticVoidPointer.h"
+#include "NativeEnclosure.h"
+
+#include "_jni/org_mozilla_jss_pkcs11_attrs_CKAClass.h"
+#include "_jni/org_mozilla_jss_pkcs11_attrs_CKAKeyType.h"
+#include "_jni/org_mozilla_jss_pkcs11_attrs_CKAUsage.h"
+#include "_jni/org_mozilla_jss_pkcs11_attrs_CKAValueLen.h"
+
+static const CK_BBOOL JSS_CK_TRUE = CK_TRUE;
+
+PRStatus
+JSS_PK11_WrapAttribute(JNIEnv *env, jobject this, void *ptr, size_t ptr_length) {
+    jclass this_class;
+    jfieldID field_id;
+    jobject ptr_object;
+    CK_ATTRIBUTE_PTR attr = calloc(1, sizeof(CK_ATTRIBUTE));
+
+    PR_ASSERT(env != NULL && this != NULL && attr != NULL);
+
+    this_class = (*env)->GetObjectClass(env, this);
+    if (this_class == NULL) {
+        goto failure;
+    }
+
+    field_id = (*env)->GetFieldID(env, this_class, "type", "J");
+    if (field_id == NULL) {
+        goto failure;
+    }
+
+    attr->type = (CK_ULONG)((*env)->GetLongField(env, this, field_id));
+    attr->pValue = ptr;
+    attr->ulValueLen = ptr_length;
+
+    ptr_object = JSS_PR_wrapStaticVoidPointer(env, (void **)&attr);
+    if (ptr_object == NULL) {
+        goto failure;
+    }
+
+    if (JSS_PR_StoreNativeEnclosure(env, this, ptr_object, sizeof(CK_ATTRIBUTE)) != PR_SUCCESS) {
+        goto failure;
+    }
+
+    return PR_SUCCESS;
+
+failure:
+    memset(attr, 0, sizeof(CK_ATTRIBUTE));
+    free(attr);
+    return PR_FAILURE;
+}
+
+PRStatus
+JSS_PK11_UnwrapAttribute(JNIEnv *env, jobject this, CK_ATTRIBUTE_PTR *attr) {
+    jobject ptr_obj;
+    jlong size = 0;
+
+    if (JSS_PR_LoadNativeEnclosure(env, this, &ptr_obj, &size) != PR_SUCCESS) {
+        goto failure;
+    }
+
+    if (JSS_PR_getStaticVoidRef(env, ptr_obj, (void **)attr) != PR_SUCCESS || attr == NULL) {
+        goto failure;
+    }
+
+    if (size != sizeof(CK_ATTRIBUTE)) {
+        goto failure;
+    }
+
+    return PR_SUCCESS;
+failure:
+    *attr = NULL;
```

Comment:
If `attr` is NULL, this code might cause segfault.

Reply:
`attr` should be non-null by calling conventions, but I can add a `PR_ASSERT(...)` at the top of the method if you'd like.

Reply:
The code in line 66 checks for null `attr`, so if that is a possibility we should check for null here too. But if that's not possible maybe the code in line 66 can be simplified?

Reply:
Ah that was a mistake. That should be `*attr == NULL`. I want to check that the returned attribute was non-NULL, not that the address space for the returned attribute was non-NULL.